distributed/cluster.py

```diff
@@ -2,65 +2,196 @@
 
 import paramiko
 from time import sleep
+import socket
+import os
 
-from toolz import assoc
+from toolz import merge
 
+# These are handy for creating colorful terminal output to enhance readability
+# of the output generated by dcluster.
+class bcolors:
+    HEADER = '\033[95m'
+    OKBLUE = '\033[94m'
+    OKGREEN = '\033[92m'
+    WARNING = '\033[93m'
+    FAIL = '\033[91m'
+    ENDC = '\033[0m'
+    BOLD = '\033[1m'
+    UNDERLINE = '\033[4m'
+
+def start_center(logdir, center_addr, center_port):
 
-def start_center(addr):
     ssh = paramiko.SSHClient()
     ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
-    ssh.connect(addr)
+    ssh.connect(center_addr, timeout = 20)
 
-    channel = ssh.invoke_shell()
-    channel.settimeout(20)
-    sleep(0.1)
-    channel.send('dcenter --host %s\n' % addr)
-    channel.recv(10000)
+    cmd = 'dcenter --host {addr} --port {port}'.format(
+        addr = center_addr, port = center_port, logdir = logdir)
 
-    return {'client': ssh, 'channel': channel}
+    # Optionally re-direct stdout and stderr to a logfile
+    if logdir is not None:
+        cmd = 'mkdir -p {logdir} && ' + cmd
+        cmd += '&> {logdir}/dcenter_{addr}:{port}.log\n'.format(
+            addr = center_addr, port = center_port, logdir = logdir)
 
+    # Run the command
+    stdin, stdout, stderr = ssh.exec_command(cmd, get_pty=True)
+    return {'address': center_addr, 'port': center_port, 'client': ssh, 'stdout': stdout, 'stderr': stderr}
 
-def start_worker(center_addr, addr):
-    ssh = paramiko.SSHClient()
-    ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
-    ssh.connect(addr)
 
-    channel = ssh.invoke_shell()
-    channel.settimeout(20)
-    sleep(0.1)
-    channel.send('dworker %s:8787\n --host %s' % (center_addr, addr))
-    channel.recv(10000)
+def start_worker(logdir, center_addr, center_port, worker_addr, workers_per_node, cpus_per_worker):
 
-    return {'client': ssh, 'channel': channel}
+    ssh = paramiko.SSHClient()
+    ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+    ssh.connect(worker_addr, timeout = 20)
+
+    # Pick a random port for the worker.  This prevents contention over a single port,
+    # which may occasionally cause workers to fail to register with the center node.
+    import random
+    worker_port = random.randint(10000, 20000)
+
+    if cpus_per_worker is None:
+        cmd = 'dworker {center_addr}:{center_port} --host {worker_addr} --port {worker_port}'.format(
+            center_addr = center_addr, center_port = center_port,
+            worker_addr = worker_addr, worker_port = worker_port,
+            logdir = logdir)
+    else:
+        cmd = 'dworker {center_addr}:{center_port} --host {worker_addr} --port {worker_port} --ncores {ncpu}\n'.format(
+            center_addr = center_addr, center_port = center_port,
+            worker_addr = worker_addr, worker_port = worker_port,
+            ncpu = cpus_per_worker, logdir = logdir)
```

Comment:
Perhaps avoid duplicated code by just adding the `--ncores` keyword in a smaller if branch